src/dmd/cparse.d

```diff
@@ -2266,6 +2267,38 @@ final class CParser(AST) : Parser!AST
                     break;
                 }
 
+                case TOK.typeof_:
+                {
+                    nextToken();
+                    check(TOK.leftParenthesis);
+
+                    void typeofExpr()
+                    {
+                        auto e = cparseExpression();
+                        auto ie = e.isIdentifierExp();
+                        auto td = !ie ? null : lookupTypedef(ie.ident);
+                        if(td && *td)
+                            t = *td;
+                        else
+                            t = new AST.TypeTypeof(token.loc, e);
+                    }
+
+                    auto tk = &token;
+                    if (isPrimaryExpression(tk))
+                        typeofExpr();
+                    // must be checked between primary/assignment lest type names instead parse as symbols
+                    else if (isTypeName(tk))
+                        t = cparseTypeName();
+                    else if (isAssignmentExpression(tk))
+                        typeofExpr();
+                    else
+                        error("`typeof` operator expects an expression or type name");
+
+                    check(TOK.rightParenthesis);
```

Comment:
It would be better to first peek `TOK.rightParenthesis`, then issue error and skip until the closing parenthesis is found, to avoid cascading errors in something like:
```
typeof(1+;2/;3*;);
```